### DISCUSSION TITLE
Cross language LTO

### DISCUSSION THREAD

**Discussion by ghost** [axboe/liburing#923] [Q&A]
https://github.com/axboe/liburing/blob/660cefae669c985aeaeaf50e45b635777ac40edc/src/Makefile#L48
llvm ThinLTO prevented until remove `-ffreestanding` `-fno-builtin`
What are they `-ffreestanding` `-fno-builtin`
Can i remove them safely
Can i enable LTO without changing Makefile s

**Comment by ghost** [comment]
```rust
Command::new("./configure")
        .arg("--cc=clang")
        .arg("--cxx=clang++")
        .current_dir(&build)
        .status()
        .unwrap();
    Command::new("make")
        .arg("-C")
        .arg(&build)
        .arg("CFLAGS=-flto=thin -g -O3 -Wall -Wextra -fno-stack-protector")
        .arg("CXXFLAGS=-flto=thin")
        .arg("V=1")
        .status()
        .unwrap();
```

**Comment by ghost** [comment]
```toml
[profile.release]
lto = "thin"
codegen-units = 1
```

**Comment by ghost** [comment]
`RUSTFLAGS="-Clinker-plugin-lto -Clinker=clang -Clink-arg=-fuse-ld=lld" cargo build --release -vv`

**Comment by ghost** [comment]
without ThinLTO
```
000000000004d410 <test_func_zzwxh>:
   4d410:	e9 8b 00 00 00       	jmp    4d4a0 <io_uring_prep_nop>
   4d415:	cc                   	int3
   4d416:	cc                   	int3
   4d417:	cc                   	int3
   4d418:	cc                   	int3
   4d419:	cc                   	int3
   4d41a:	cc                   	int3
   4d41b:	cc                   	int3
   4d41c:	cc                   	int3
   4d41d:	cc                   	int3
   4d41e:	cc                   	int3
   4d41f:	cc                   	int3
```

**Comment by ghost** [comment]
with ThinLTO
```
000000000004d410 <test_func_zzwxh>:
   4d410:	48 b8 00 00 00 00 ff 	movabs $0xffffffff00000000,%rax
   4d417:	ff ff ff 
   4d41a:	48 89 07             	mov    %rax,(%rdi)
   4d41d:	0f 57 c0             	xorps  %xmm0,%xmm0
   4d420:	0f 11 47 28          	movups %xmm0,0x28(%rdi)
   4d424:	48 c7 47 38 00 00 00 	movq   $0x0,0x38(%rdi)
   4d42b:	00 
   4d42c:	0f 11 47 08          	movups %xmm0,0x8(%rdi)
   4d430:	48 c7 47 18 00 00 00 	movq   $0x0,0x18(%rdi)
   4d437:	00 
   4d438:	c3                   	ret
   4d439:	cc                   	int3
   4d43a:	cc                   	int3
   4d43b:	cc                   	int3
   4d43c:	cc                   	int3
   4d43d:	cc                   	int3
   4d43e:	cc                   	int3
   4d43f:	cc                   	int3
```

**Comment by ghost** [comment]
```rust
#[inline(never)]
#[no_mangle]
pub fn test_func_zzwxh(sqe:&mut io_uring_sqe) {
    unsafe {
        io_uring_prep_nop(sqe);
    }
}
```

**Comment by ghost** [comment]
@ammarfaizi2 @axboe

**Comment by ammarfaizi2** [comment]
> What are they `-ffreestanding` `-fno-builtin`

Those flags are needed to compile liburing nolibc.

> Can i remove them safely

Just do `./configure --use-libc` before `make`.

> Can i enable LTO without changing Makefile s

You probably need to alter the `LDFLAGS`.

**Reply by ghost** [reply]
unknown option --use-libc

**Reply by ghost** [reply]
Can you explain why thinLTO get prevented?

**Reply by ghost** [reply]
use libc is not released
Can you release a new version

**Reply by ghost** [reply]
Is it safe and stable to use the mainbranch?

**Reply by ammarfaizi2** [reply]
> Can you explain why thinLTO get prevented?

I said `--use-libc` will remove those options. But those options don't
bother the LTO. I can do LTO without removing those options anyway.

`-ffreestanding` and `-fno-builtin` don't prevent LTO.

### test_lto.c
```c

#include "liburing.h"

void test_lto(struct io_uring_sqe *sqe)
{
	io_uring_prep_nop(sqe);
}
```

### Compile liburing
```
./configure --cc=clang --cxx=clang++;

make CFLAGS="-flto -g -O3" LDFLAGS="-flto" V=1 -j8;
```

### Compile the test program
```
clang -O3 -Isrc/include -flto test_lto.c -o test_lto.so -shared src/liburing-ffi.a;

objdump -d test_lto.so;
```

```asm
00000000000010fc <test_lto>:
    10fc:	48 b8 00 00 00 00 ff 	movabs $0xffffffff00000000,%rax
    1103:	ff ff ff 
    1106:	48 89 07             	mov    %rax,(%rdi)
    1109:	0f 57 c0             	xorps  %xmm0,%xmm0
    110c:	0f 11 47 08          	movups %xmm0,0x8(%rdi)
    1110:	31 c0                	xor    %eax,%eax
    1112:	48 89 47 18          	mov    %rax,0x18(%rdi)
    1116:	0f 11 47 28          	movups %xmm0,0x28(%rdi)
    111a:	48 89 47 38          	mov    %rax,0x38(%rdi)
    111e:	c3                   	ret  
```

**Reply by ammarfaizi2** [reply]
> Is it safe and stable to use the mainbranch?

The master branch tends to be not stable. Unreleased things may come and
go. But it's not that dangerous to use liburing master.

**Reply by alviroiskandar** [reply]
@ammarfaizi2 since you include liburing.h without defining IOURINGINLINE, it will be static inline, so it doesn't prove you can use LTO here.

**Reply by alviroiskandar** [reply]
@ammarfaizi2 try this and see if the LTO works.
```
struct io_uring_sqe;
extern void io_uring_prep_nop(struct io_uring_sqe *sqe);

void test_lto(struct io_uring_sqe *sqe)
{
	io_uring_prep_nop(sqe);
}
```

**Reply by alviroiskandar** [reply]
Ah yeah, you're right. Now tested with my suggested test_lto code. It still can do LTO.

```asm
0000000000003120 <test_lto>:
    3120:       48 b8 00 00 00 00 ff    movabs $0xffffffff00000000,%rax
    3127:       ff ff ff 
    312a:       48 89 07                mov    %rax,(%rdi)
    312d:       0f 57 c0                xorps  %xmm0,%xmm0
    3130:       0f 11 47 28             movups %xmm0,0x28(%rdi)
    3134:       48 c7 47 38 00 00 00    movq   $0x0,0x38(%rdi)
    313b:       00 
    313c:       0f 11 47 08             movups %xmm0,0x8(%rdi)
    3140:       48 c7 47 18 00 00 00    movq   $0x0,0x18(%rdi)
    3147:       00 
    3148:       c3                      ret    
    3149:       0f 1f 80 00 00 00 00    nopl   0x0(%rax)
```

**Reply by ammarfaizi2** [reply]
> Ah yeah, you're right. Now tested with your suggested test_lto code. It still can do LTO.

Oh yeah, it was a mistake. But still, LTO works. The fact
that you can inline an "extern" function is the proof that
LTO works.

**Reply by ghost** [reply]
I test with rust, lto is blocked until use-libc is enabled

**Comment by ghost** [comment]
Although you believe that `-fno-builtin` does not prevent LTO, I found that it does indeed prevent inlining, according to the LLVM optimization log. @ammarfaizi2 @alviroiskandar

**Comment by ghost** [comment]
https://discord.com/channels/636084430946959380/636732781086638081/1149426538811510924

**Comment by ghost** [comment]
```
remark: <unknown>:0:0: 'io_uring_queue_init' not inlined into 'test_inline1' because it should never be inlined (cost=never): conflicting attributes [-Rpass-missed=inline]
remark: <unknown>:0:0: 'io_uring_get_sqe' not inlined into 'test_inline1' because it should never be inlined (cost=never): conflicting attributes [-Rpass-missed=inline]
remark: <unknown>:0:0: '_ZN4core9panicking5panic17hee69a8315e4031d6E' not inlined into 'test_inline1' because it should never be inlined (cost=never): noinline function attribute [-Rpass-missed=inline]
remark: <unknown>:0:0: 'io_uring_prep_nop' not inlined into 'test_inline1' because it should never be inlined (cost=never): conflicting attributes [-Rpass-missed=inline]
remark: <unknown>:0:0: 'io_uring_sqe_set_data64' not inlined into 'test_inline1' because it should never be inlined (cost=never): conflicting attributes [-Rpass-missed=inline]
remark: <unknown>:0:0: 'io_uring_submit' not inlined into 'test_inline1' because it should never be inlined (cost=never): conflicting attributes [-Rpass-missed=inline]
remark: <unknown>:0:0: '_ZN4core9panicking5panic17hee69a8315e4031d6E' not inlined into 'test_inline1' because it should never be inlined (cost=never): noinline function attribute [-Rpass-missed=inline]
remark: <unknown>:0:0: 'io_uring_wait_cqe' not inlined into 'test_inline1' because it should never be inlined (cost=never): conflicting attributes [-Rpass-missed=inline]
remark: <unknown>:0:0: '_ZN4core9panicking5panic17hee69a8315e4031d6E' not inlined into 'test_inline1' because it should never be inlined (cost=never): noinline function attribute [-Rpass-missed=inline]
remark: <unknown>:0:0: 'io_uring_cqe_get_data64' not inlined into 'test_inline1' because it should never be inlined (cost=never): conflicting attributes [-Rpass-missed=inline]
remark: <unknown>:0:0: '_ZN4core9panicking5panic17hee69a8315e4031d6E' not inlined into 'test_inline1' because it should never be inlined (cost=never): noinline function attribute [-Rpass-missed=inline]
remark: <unknown>:0:0: 'io_uring_cqe_seen' not inlined into 'test_inline1' because it should never be inlined (cost=never): conflicting attributes [-Rpass-missed=inline]
remark: <unknown>:0:0: 'io_uring_queue_exit' not inlined into 'test_inline1' because it should never be inlined (cost=never): conflicting attributes [-Rpass-missed=inline]
```

**Comment by ghost** [comment]
`because it should never be inlined (cost=never): conflicting attributes`
When i add no-builtin and freestanding flags, llvm says `conflicting attributes`

**Comment by ghost** [comment]
https://github.com/zzwxh/ltotest-liburing

**Comment by ghost** [comment]
https://reviews.llvm.org/D74162

**Comment by ghost** [comment]
> Here I have allowed inlining to proceed if the callee's nobuiltins are a
subset of the caller's nobuiltins, but not in the reverse case, which
should be conservatively correct.

**Comment by cmazakas** [comment]
Hmm, you know, the more I think about this, I'm almost wondering if it's just outright better to transliterate liburing to Rust and maintain it as a separate fork. This, of course, brings about the possibility of some impl divergence and lag but seeing the contortions I had to go through and what zzwxh is going through, a native impl might be more appropriate.

Fundamentally, liburing is just a userspace wrapper around io_uring. I don't see any inherent reason why we couldn't just translate this code.

**Reply by ghost** [reply]
[liburing-lto ](https://crates.io/crates/liburing-lto)

**Reply by ghost** [reply]
This issue on master branch has been resolved.
I'm tring to bring lto to liburing2.4

**Reply by ghost** [reply]
Actually if liburing releases 2.5, I will stop all these work

**Reply by ghost** [reply]
These are not really issues. The cost of function calls is very low, and using the master branch usually doesn't pose any problems; it's just annoying.

**Reply by ghost** [reply]
I've already tried rewriting liburing in Rust, which is highly demanding in terms of concurrency safety. Some parts of the liburing codebase are overly complex, especially when it comes to function naming and unclear control flow. There are some ugly patches for compatibility, like LIBURING_UDATA_TIMEOUT. The abstraction level is blurry and exposes a lot of unnecessary details, unlike the Microsoft IORING API, which is simpler. However, the advantage of liburing is that it's a popular library, not prone to becoming obsolete, and its maintainers are skilled. In conclusion, splitting the community makes no sense; we should rather focus on improving liburing.

**Reply by cmazakas** [reply]
Liburing is  single threaded lib. What concurrency?

**Reply by ghost** [reply]
IORING_SETUP_SQPOLL

**Reply by ghost** [reply]
In fact, due to the presence of interrupts, the single-threaded environment you perceive does not exist. Consider why you can use `io_uring_peek_cqe` for polling. The `sqpoll` mode absolutely requires synchronization, but that doesn't mean the interrupt mode doesn't need it or can take shortcuts.
@cmazakas

**Reply by cmazakas** [reply]
All of io_uring requires synchronization with the kernel but that really doesn't need to bleed out into the Rust API. The Rust API, as far as I can tell, wouldn't need Send or Sync at all and in fact _should_ be !Send and !Sync.

**Reply by redbaron** [reply]
There is already https://github.com/tokio-rs/io-uring/ which is a low level binding to kernel uring, similar to what liburing is, but native and idiomatic Rust. It has no dependencies  other than `libc`

**Reply by cmazakas** [reply]
```rust
    let read_e = opcode::Read::new(types::Fd(fd.as_raw_fd()), buf.as_mut_ptr(), buf.len() as _)
        .build()
        .user_data(0x42);

    // Note that the developer needs to ensure
    // that the entry pushed into submission queue is valid (e.g. fd, buffer).
    unsafe {
        ring.submission()
            .push(&read_e)
            .expect("submission queue is full");
    }
```
look how they massacred my boy!

**Reply by ghost** [reply]
I tried rewriting liburing again, and encountered some issues, especially within sq_thread_idle where I repeatedly received IORING_SQ_NEED_WAKEUP. I believe this is due to incorrect synchronization, but I'm not sure where to start troubleshooting.
@cmazakas

**Reply by ghost** [reply]
It is also possible that it is a kernel bug or I misunderstood the SQ_POLL mode.

**Comment by lorislibralato** [comment]
Hi @aaaaaaaaaaasjksj,
I'm trying to link statically liburing in rust, can you show me the right build setup for LTO?

**Reply by ghost** [reply]
https://crates.io/crates/liburing2-sys
It works fine, but I won't maintain it anymore
@lorislibralato
